[스프린트 2주차] 미션 제출합니다.

gxxrxn opened the pull request:
이번 스프린트를 통해 forEach, map 등 Array의 메소드 사용에 익숙해진 것 같습니다.
감사합니다.

gxxrxn committed:
docs: 요구사항 목록 작성

gxxrxn committed:
feat: customReduce 기능 구현

- domain/array.js에 customReduce 구현
- log.js 모듈 export 오류 수정

gxxrxn committed:
docs: customReduce 구현에 따른 요구사항 목록 업데이트

gxxrxn committed:
feat: add, count 함수 구현

- operations.js에 customReduce를 사용하는 add, count 함수 구현

gxxrxn committed:
feat: Commander 명령어 파싱 기능

- Commander 클래스의 parse 함수 구현
- 명령어를 공백에 따라 split
- add 명령인 경우 값들을 숫자로 변환
- Commander의 명령 옵션으로 add, count를 가지도록 구현

gxxrxn committed:
feat: Commander 클래스 exec 함수 구현

- 명령어에 따라 명령 옵션 함수를 실행하는 exec 함수 구현

gxxrxn committed:
feat: parse와 exec을 실행하는 run 함수 구현

- parse, exec을 실행하여 명령어와 실행결과를 전달

gxxrxn committed:
docs: 기능 구현에 따른 요구사항 목록 업데이트

gxxrxn committed:
feat: Log 클래스 구현

- 태그와 값을 전달받아 결과를 출력

gxxrxn committed:
docs: Log 구현에 따른 요구사항 목록 업데이트

gxxrxn committed:
feat: testcase에 따라 commander 실행하는 main 함수 구현

gxxrxn committed:
docs: 기능 구현에 따른 요구사항 목록 업데이트

- 입력 및 예외처리 관련 요구사항 구현 완료
- main 함수 구현 완료

SingTheCode commented on sprint-2/domain/array.js:
```diff
@@ -1,5 +1,18 @@
+  const length = array.length;
+
+  let index = 0;
+  let accumulator = initialValue ?? array[index++];
```
Nullish Coalescing Operator라는 것을 처음 알게되었습니다. 감사합니다!

2yunseong replied:
reduce 설명에 따라 코드를 깔끔하게 작성해주신 것 같아요. 인덱스 처리도 저는 개별로 했는데 한줄의 코드에 담은 점이 깔끔하네요

SingTheCode commented on sprint-2/domain/array.js:
```diff
@@ -1,5 +1,18 @@
+  const array = Object(this);
+  const length = array.length;
+
+  let index = 0;
+  let accumulator = initialValue ?? array[index++];
+  let currentValue = array[index];
```
initialValue의 유무를 이렇게 처리하셨군요!! 코드 길이도 짧아졌을뿐더러 가독성도 good 인 것 같습니다!!

SingTheCode commented on sprint-2/main.js:
```diff
@@ -1,4 +1,15 @@
+  const log = new Log();
+  const testCase = ["count 1 2 3 4 5", "add 35 23"];
+
+  testCase.forEach((tc) => {
+    const [option, result] = commander.run(tc);
+    log.result(option, result);
+  });
```
원하는 테스트케이스만 골라서 메서드를 수행할 상황을 생각하셔서 만든 코드라면 정말 좋은 방법인 것 같습니다! 다만 모든 testCase에 대해 로직을 실행해야된다면 run함수가 밖으로 나오지 않고 Class 안에서 처리되는건 어떨까요??

gxxrxn replied:
문제 요구사항에서 입력이 문자열이라 commander가 하나의 문자열 입력만 처리하도록 구현했습니다..!!
배열의 모든 테스트케이스들을 한번에 처리하는 로직은 commander 입장에서는 불필요한 것 같아서요..
(사실 테스트케이스들을 배열로 만든것도 제가 임의로 만든거라..)

run 함수가 클래스 내부에서 어떻게 처리되면 좋을 것 같다는 말씀이실까요!?!
생성자로 입력이 전달되고 생성자에서 실행되도록..?
제가 잘 이해한건지 모르겠네요..ㅎㅎ

SingTheCode replied:
넵!! 말씀하신 부분이 맞습니다!

SingTheCode commented on sprint-2/domain/commander.js:
```diff
@@ -1,5 +1,29 @@
+    this.options = { add, count };
+  }
+
+  parse(command) {
+    let [option, ...array] = command.split(" ");
+
+    if (option == "add") {
+      array = array.map((x) => Number(x));
+    }
+
+    return [option, array];
+  }
```
한 메서드는 한 역할만 수행한다는 규칙이 잘 지켜진 것 같아요! 코드의 규모가 커질수록 이렇게 함수를 분리하는 것이 필요할 것 같습니다!

SingTheCode reviewed (commented):
요구사항을 입맛대로 바꿔서 체크리스트를 깔끔하게 작성하신게 인상적입니다! 파일과 메서드를 역할에 따라 분리하려는 노력을 많이 하신게 보이는 것 같아요!! 코드에 대한 제 생각을 아래에 추가로 남길게요!!
- add, count 함수를 operations.js 파일로 분리한 것이 인상적이였습니다. parse 함수를 통해 option의 case가 추가될 때마다 다른 parsing을 할 수 있게 분리 시켜놓은 것도 좋은 것 같아요. 하나 고민이 되는 건 분리를 어디까지 해야하나에 대한 고민입니다. 많이 분리시킨 코드는 오히려 따라가기가 힘들 수도 있다는 생각이 들어서 코드의 규모에 따라 분리의 정도를 조절해야 한다라고 저는 생각하는데 다른 분들은 어떤 생각이실지 궁금합니다. 사실 parse() 메서드도 따로 parse.js 파일로 분리할 수 있으니까요!

2yunseong commented on sprint-2/domain/commander.js:
```diff
@@ -1,5 +1,29 @@
+  }
+
+  parse(command) {
+    let [option, ...array] = command.split(" ");
```
고급 기능 가르쳐주셔서 감삼다 ^

2yunseong commented on sprint-2/domain/operations.js:
```diff
@@ -0,0 +1,4 @@
+const add = (array) => array.customReduce((acc, cur) => acc + cur);
+const count = (array) => array.customReduce((acc, cur) => acc + 1, 0);
```
callback 함수를 따로 분리해서 작성해주는건 어떤가요? 지금은 상관없지만, 만약 add, count 이외 다른 콜백함수를 사용해야 할 상황이 생긴다면 따로 분리해서 작성해주는게 일관성이 생기지 않을까라는 개인적인 생각입니다.

2yunseong reviewed (commented):
코드 잘 봤습니다. 수요일날 봐요 규란님

gxxrxn committed:
feat: sync-async 함수 구현

gxxrxn committed:
feat: Shop, Employee 클래스 구현

- shop의 order 메소드 구현
- employee의 makeBurger 메소드 구현

gxxrxn committed:
refactor: shop 메소드 분리

- work 메소드
  - 주문 수량이 0이 될 때까지 직원들이 버거를 만들도록 함.
- finishOrder 메소드
  - 현재 intervalID를 전달하면 interval을 종료
- announce 메소드
  - 현재까지의 주문을 만드는데 걸린 시간 orderTime과 메시지를 전달받아 출력
  - 시간이 없으면 출력하지 않음.
